backend/common/src/main/java/ai/verta/modeldb/common/db/migration/Migrator.java

```diff
@@ -22,19 +31,150 @@ public Migrator(Connection connection, String resourcesDirectory) {
   /** Assumes that the underlying database has already been created. */
   public void performMigration(RdbConfig config) throws SQLException, MigrationException {
     log.info("Starting database migration process");
-    MigrationDatastore migrationDatastore = MigrationDatastore.create(config, connection);
+    performMigration(config, null);
+  }
 
+  void performMigration(RdbConfig config, Integer desiredVersion)
+      throws SQLException, MigrationException {
+    MigrationDatastore migrationDatastore = setupDatastore(config);
+    MigrationState currentState = findCurrentState();
+    if (currentState == null) {
+      throw new IllegalStateException(
+          "The schema_migrations table contains no records. Migration process cannot start.");
+    }
+    if (currentState.isDirty()) {
+      // todo: implement dirty handling
+      throw new MigrationException(
+          "Database is in a dirty state. Bailing out until dirty handling is implemented.");
+    }
+    int versionToTarget =
+        desiredVersion != null
+            ? desiredVersion
+            : gatherMigrations(migration -> true, migration -> true).stream()
+                .map(Migration::getNumber)
+                .max(Integer::compareTo)
+                .orElse(0);
+    log.info("Starting database migration process to version: " + versionToTarget);
+
+    if (currentState.getVersion() == versionToTarget) {
+      log.info("No migrations to perform. Database is already at version " + versionToTarget);
+      return;
+    }
+    NavigableSet<Migration> migrationsToPerform =
+        gatherMigrationsToPerform(currentState, versionToTarget);
+    runMigrations(migrationDatastore, migrationsToPerform);
+  }
+
+  private NavigableSet<Migration> gatherMigrationsToPerform(
+      MigrationState currentState, int versionToTarget) throws MigrationException {
+    NavigableSet<Migration> migrationsToPerform;
+    if (currentState.getVersion() < versionToTarget) {
+      migrationsToPerform =
+          gatherMigrations(
+              Migration::isUp,
+              migration ->
+                  (migration.getNumber() > currentState.getVersion())
+                      && (migration.getNumber() <= versionToTarget));
+    } else {
+      migrationsToPerform =
+          gatherMigrations(
+                  Migration::isDown,
+                  migration ->
+                      migration.getNumber() > versionToTarget
+                          && migration.getNumber() <= currentState.getVersion())
+              .descendingSet();
```

Comment:
This is the only place I see using a method from `NavigableSet`.

Reply:
yes, but the orderedness is critical to the processing, so keeping that semantic information in the return type seems very important to me.

Reply:
changed return type to SortedSet